api/proto/config/storage.proto

```diff
@@ -193,3 +193,32 @@ message Volume {
   string displayName = 7;  // Optional friendly name echo'ed in info message
   bool clear_text = 8;  // Flag to indicate the volume encryption needed or not
 }
+
+// DiskConfigType is the desired configuration of disks
+enum DiskConfigType {
+  DISK_CONFIG_TYPE_UNSPECIFIED = 0; // no configured, do nothing
+  DISK_CONFIG_TYPE_EVEOS = 1; // the disk EVE is installed on
+  DISK_CONFIG_TYPE_PERSIST = 2; // the disk is separate persist partition, not zfs
+  DISK_CONFIG_TYPE_ZFS_ONLINE = 3; // included in zfs and online
+  DISK_CONFIG_TYPE_ZFS_OFFLINE = 4; // included in zfs and offline
+  DISK_CONFIG_TYPE_APPDIRECT = 5; // for direct assignment
+  DISK_CONFIG_TYPE_UNUSED = 6; // removed from zfs/app-direct
+}
+
+//DiskConfig describe desired configuration of disk
+message DiskConfig {
+  string disk_name = 1;
+  string old_disk_name = 2; // define old disk name here to replace it with new one defined in disk_name
+  DiskConfigType disk_config = 3;
+}
+
+//DisksConfig is a configuration of disks
+//We expect information about disks to be filled and will try to adjust disks states accordingly
+//If we want change state to online/offline we should define its state
+//If we want to add disk we should define it here and set DiskConfigType to online or offline
+//If we want to remove disk we should set its state to removed
+//If we want to replace disk we should fill old_disk_name to be replaced with disk_name
```

Comment:
Just FYI. Replace disks operation can happen in two ways. 

1) Replace disk at same place, say sdc failed, user will remove the disk and plug in new disk at same slot. Then command will be just zpool replace persist /dev/sdc
In such cases I presume old_disk_name and disk_name will be same.

2) Replace sdc with another disk say sdd. sdc is still plugged in but is in bad state. Then command is zpool replace persist /dev/sdc /dev/sdd.  In such cases disk_name=sdd and old_disk_name=sdc

Reply:
But in the case of 1, suppose tech removes /dev/sdc and then there is a power outage before tech can plug in the replacement, then what happens? Will /dev/sdd become /dev/sdc when it reboots due to the power outage? If so things will be very confused.

It would seem more robust to say disk_serial=X replaced_serial=Y to handle that case. And if we have created an alias such as "disk3" for serial Y, then one can say disk_serial=X replaced_logical=disk3.

Reply:
It depends, if the OS gives the new drive name as sdc again, then the #1 command should be run. Else #2.
But ZFS does not work on serial numbers unfortunately and it only goes by device names. But we can use serial numbers to avoid confusion.

So IMO the structure should  be.

{ 
  disk_name
  disk_serial
  old_disk_name
  old_disk_serial
}

Reply:
Maybe it will be better to have another message, say "disk_description" and store name and serial there?

Reply:
Makes sense to me.

Reply:
Done

Reply:
LGTM. Since we are describing the disk it might also makes sense to have size and type of the disk ?

Reply:
I put it into the structure to be expandable in future. Probably we will go with "label" or with "uuid", we need to have some kind of pointer onto the disk we want to touch.

Reply:
Ok it's up to you. If we have it right now then need not worry about UI integration part later.

Reply:
Can you please elaborate what you mean by size and type of the disk here? I cannot understand the reason of adding it into configuration we will receive from the controller.

Reply:
Couple of reasons I can think of.

1) Replacing a disk in ZFS pool requires a disk of same size or larger, we need to make sure user is using correct sized disks and error out if not.
2) Also, the device type too ie we want all SSDs or NVMe or HDD drives in a zpool, we do not want a mixture of drives. At least warn if not error out.

Reply:
Correct, we should check it and respond internally, but I do not think that we should ask user to put this information in config, config should contain entities that we can configure

Reply:
That works too. I was thinking since UI is sending this information they might as well send the disk size and type. But if they are not needed then I am fine with it.

Reply:
@zedi-pramodh I think the controller and/or user needs to do checks around size, thus reporting that in the info messages sent from EVE makes sense. But it doesn't make sense to but a size in this config API.